File: AP-Comp-Codes/04242023 Data Science/Lesson 3/QuickSort.py
import random, time
import numpy as np


iterations = 0

def swap(array, i, j):
    tmp = array[i]
    array[i] = array[j]
    array[j] = tmp

def partition(array, start, end):
    """ quicksort partitioning, using end """
    pivot = array[end]
    L = start
    R = end
    while L < R:
        while array[L] < pivot:
            L += 1
        while array[R] > pivot:
            R -= 1
        swap(array, L, R)
        # avoid hanging on the same numbers
        if ( array[L] == array[R] ):
            L += 1
    return R

def _quicksort(array, start, end):
    """ Recursive quicksort function """
    global iterations
    iterations += 1

    print("Recursive Sorting. Iterations:", iterations)

    if start < end:
        split = partition(array, start, end)
        _quicksort(array, start, split-1)
        _quicksort(array, split+1, end)

def quicksort(array):
    global iterations
    iterations = 0
    _quicksort(array, 0, len(array)-1)
    return iterations

if __name__ == "__main__":
    # array = [12, 5, 11, 6, -3, -4, -11, 6, 3, 4, 1, -2]
    arraySize = int(input("Enter number of elements:\n$ "))
    array = [random.randint(-1000, 1000) for _ in range(arraySize)]
    npArray = np.array(array)
    print(npArray)

    input()

    start_time = time.time()

    quicksort(array)
    
    end_time = time.time()

    elapsed_time = end_time - start_time

    print(array)
    print("Iterations taken:", iterations)
    print("Array size:      ", arraySize)
    print("Time taken:      ", elapsed_time)

File: AP-Comp-Codes/04242023 Data Science/C++ Playground/QuickSort.cpp
#include <iostream>
#include <algorithm>
#include "QuickSort.h"


QuickSort::QuickSort(std::vector<int> listToSort) {
    // constructor
    sortingList = listToSort;
    isSorted = false;
}

void QuickSort::swap(std::vector<int>& array, int i, int j) {
    int tmp = array[i];
    array[i] = array[j];
    array[j] = tmp;
}

int QuickSort::partition(std::vector<int>& array, int start, int end) {
    // quicksort partitioning, using end

    // variable setup
    int pivot = array[end];
    int L = start + 0;
    int R = end + 0;

    // main while loop
    while (L < R) {
        // increment left
        while (array[L] < pivot) {
            L++;
        }
        // increment right
        while (array[R] > pivot) {
            R--;
        }
        // swap values
        swap(array, L, R);
        // avoid hanging on the same numbers
        if (array[L] == array[R]) {
            L++;
        }
    }
    // return right
    return R;
}

void QuickSort::_quicksort(std::vector<int>& array, int start, int end) {
    // recursive quicksort function
    //increment iteration count
    iterations++;

    std::cout << "Recursive sorting. Iterations: " << iterations << "\n"; 

    // run sorting if sorting hasn't reached the end
    if (start < end) {
        // get split
        int split = partition(array, start, end);
        _quicksort(array, start, split-1);
        _quicksort(array, split+1, end);
    }
}

void QuickSort::sortArray() {
    // equivalent to python's "quicksort" function for initial sorting
    iterations = 0;

    std::cout << "Sorting array\n";

    // call recursive sorting function
    _quicksort(sortingList, 0, sortingList.size()-1);
    // set sorted bool val
    isSorted = true;
}

int QuickSort::getIterations() {
    return iterations;
}

std::vector<int> QuickSort::getSortedList() {
    return sortingList;
}